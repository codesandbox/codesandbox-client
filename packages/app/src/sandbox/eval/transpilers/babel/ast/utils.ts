import * as meriyah from 'meriyah';
import * as astring from 'astring';
import { customGenerator } from './generator';

export interface ESTreeAST {
  isDirty: boolean;
  program: meriyah.ESTree.Program;
}

export function parseModule(code: string): ESTreeAST {
  return {
    isDirty: false,
    program: meriyah.parseModule(code, {
      next: true,
      raw: true,
      jsx: true,
    }),
  };
}

export function generateCode(ast: ESTreeAST) {
  const finalCode = astring.generate(ast.program as any, {
    generator: customGenerator,
  });

  return `"use strict";\n${finalCode}`;
}
