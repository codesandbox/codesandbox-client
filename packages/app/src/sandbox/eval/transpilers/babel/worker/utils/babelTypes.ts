let types: unknown = null;

/**
 * Load babel types if the fs is needed, it greatly speeds up compilation of
 * custom plugins and presets, but we don't want to include it in the main bundle.
 */
export function loadBabelTypes(): Promise<void> {
  return import('@babel/types').then(typesResult => {
    types = typesResult;
  });
}

export function getBabelTypes() {
  if (!types) {
    throw new Error("Babel types aren't loaded");
  }

  return types;
}
