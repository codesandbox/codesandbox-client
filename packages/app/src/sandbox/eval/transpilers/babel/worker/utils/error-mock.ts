/* eslint-disable no-extend-native, no-useless-computed-key */
const getFileName = () => '/index.js';
const stackSplit = { getFileName };

let hasStack = false;

function checkHasStack() {
  if (hasStack) {
    return;
  }

  const origPrepareStackTrace = Error.prepareStackTrace;

  // eslint-disable-next-line
  Error.prepareStackTrace = function (_, stack) {
    return stack;
  };

  const stack = new Error().stack;
  Error.prepareStackTrace = origPrepareStackTrace;

  // @ts-ignore
  hasStack = typeof stack[0].getFileName === 'function';
}

checkHasStack();

class NewError extends Error {
  constructor(n: string) {
    super(n);

    // @ts-ignore
    this.stack = {
      [2]: stackSplit,
      split() {
        return [stackSplit, stackSplit, stackSplit, stackSplit];
      },
      replace() {
        return '';
      },
    };
  }
}

/**
 * Babel expects to be able to do this:
 *
 * ```js
 * const error = new Error();
 * const fileName = error.stack.split()[2].getFileName();
 * ```
 *
 * This works, but only in the V8 engine, so Firefox and Safari will break.
 * This function will hardcode the functionality.
 */
export function installErrorMock() {
  if (!hasStack) {
    // @ts-ignore
    self.Error = NewError;

    hasStack = true;
  }
}
