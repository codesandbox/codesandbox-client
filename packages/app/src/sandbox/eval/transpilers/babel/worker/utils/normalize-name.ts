export function normalizePluginName(name: string) {
  return name.replace('babel-plugin-', '').replace('@babel/plugin-', '');
}

export function normalizePresetName(name: string) {
  return name.replace('babel-preset-', '').replace('@babel/preset-', '');
}
