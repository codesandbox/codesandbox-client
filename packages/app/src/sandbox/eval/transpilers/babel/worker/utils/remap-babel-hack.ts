export function remapBabelHack() {
  if (self.module.exports.availablePlugins) {
    // Okay, Babel thought it was in a commonjs env and now put everything on
    // self.exports. We need to remap it.

    // @ts-ignore
    self.Babel = self.module.exports;

    // Reset module for next time Babel loads
    // @ts-ignore
    self.module = { exports: {} };
    const { exports } = self.module;
    self.exports = exports;
  }
}
