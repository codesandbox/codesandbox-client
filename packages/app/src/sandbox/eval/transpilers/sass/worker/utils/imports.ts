const lineRegex = /@import\s*['|"|`]([^"|'|`]*)['|"|`]/;

export function getImports(code: string): string[] {
  const lines = code.split('\n');
  const result = new Set();
  lines.forEach(line => {
    const match = line.match(lineRegex);
    if (match && match[1]) {
      result.add(match[1]);
    }
  });

  return Array.from(result) as string[];
}
