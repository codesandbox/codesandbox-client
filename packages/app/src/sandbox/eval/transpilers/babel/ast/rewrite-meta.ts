// This code is written to be performant, that's why we opted to ignore these linting issues
/* eslint-disable no-loop-func, no-continue */
import * as meriyah from 'meriyah';
import { walk } from 'estree-walker';

import { Syntax as n } from './syntax';
import { ESTreeAST } from './utils';

export interface IESModuleMeta {
  url: string;
}

const CSB_IMPORT_META_NAME = '$csb__import_meta';
export function rewriteImportMeta(ast: ESTreeAST, meta: IESModuleMeta): void {
  let hasImportMeta = false;
  walk(ast.program, {
    enter(node: meriyah.ESTree.MemberExpression) {
      if (node.type === n.MemberExpression) {
        if (
          node.object.type === n.MetaProperty &&
          node.object.meta.name === 'import'
        ) {
          node.object = {
            type: n.Identifier,
            name: CSB_IMPORT_META_NAME,
          };

          hasImportMeta = true;

          this.skip();
        }
      }
    },
  });

  if (hasImportMeta) {
    ast.program.body.unshift({
      type: n.VariableDeclaration,
      kind: 'var',
      declarations: [
        {
          type: n.VariableDeclarator,
          id: {
            type: n.Identifier,
            name: CSB_IMPORT_META_NAME,
          },
          init: {
            type: n.ObjectExpression,
            properties: [
              {
                type: n.Property,
                kind: 'init',
                computed: false,
                shorthand: false,
                method: false,
                key: {
                  type: n.Identifier,
                  name: 'url',
                },
                value: {
                  type: n.Literal,
                  value: meta.url,
                  raw: JSON.stringify(meta.url),
                },
              },
            ],
          },
        },
      ],
    });
  }
}
