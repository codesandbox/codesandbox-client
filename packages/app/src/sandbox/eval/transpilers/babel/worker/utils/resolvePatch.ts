import * as resolve from 'browser-resolve';
import { packageFilter } from 'sandbox/eval/utils/resolve-utils';

/**
 * Patch 'resolve' to configure it to resolve esmodules by default. babel-plugin-macros goes
 * for the commonjs version, which we don't download anymore by default.
 */
export function patchedResolve() {
  const handler = {
    get(target, prop) {
      if (prop === 'sync') {
        return (p, options) => target.sync(p, { ...options, packageFilter });
      }

      return target[prop];
    },
  };
  return new Proxy(resolve, handler);
}
