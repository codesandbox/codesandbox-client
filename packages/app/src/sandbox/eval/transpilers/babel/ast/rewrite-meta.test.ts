import { rewriteImportMeta } from './rewrite-meta';
import { generateCode, parseModule } from './utils';

describe('Rewrite import.meta', () => {
  it('Can detect and rewrite import.meta to $csb__import_meta', () => {
    const code = `const a = import.meta.url;`;
    const ast = parseModule(code);
    rewriteImportMeta(ast, {
      url: 'https://csb.io/index.js',
    });
    const result = generateCode(ast);
    expect(result).toMatchSnapshot();
  });

  it('Should not add a global $csb__import_meta when there is no need for it', () => {
    const code = `const a = "hello";`;
    const ast = parseModule(code);
    rewriteImportMeta(ast, {
      url: 'https://csb.io/index.js',
    });
    const result = generateCode(ast);
    expect(result).toMatchSnapshot();
  });
});
