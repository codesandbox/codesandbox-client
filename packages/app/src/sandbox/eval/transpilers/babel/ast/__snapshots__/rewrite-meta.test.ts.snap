// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rewrite import.meta Can detect and rewrite import.meta to $csb__import_meta 1`] = `
"\\"use strict\\";
var $csb__import_meta = {
  url: \\"https://csb.io/index.js\\"
};
const a = $csb__import_meta.url;
"
`;

exports[`Rewrite import.meta Should not add a global $csb__import_meta when there is no need for it 1`] = `
"\\"use strict\\";
const a = \\"hello\\";
"
`;
